docs/reference/modules/coordination.asciidoc

```diff
@@ -0,0 +1,461 @@
+|0                                        |1
+|**1**                                    |**3**
+|2                                        |5
+|3                                        |7
+|====================================================================
+
+It is permissible, but not recommended, to set
+`cluster.master_nodes_failure_tolerance` too low for your cluster or, put
+differently, to have more master-eligible nodes than the minimum configuration
+size. To do so is _safe_ in the sense that the cluster will not suffer from a
+split-brain however this setting is configured, but if
+`cluster.master_nodes_failure_tolerance` is too low then your cluster may not
+tolerate as many failures as expected.
+
+[float]
+==== Even numbers of master-eligible nodes
+
+There should normally be an odd number of master-eligible nodes in a cluster.
+If there is an even number then Elasticsearch will put all but one of them into
+the configuration to ensure that the configuration has an odd size. This does
+not decrease the failure-tolerance of the cluster, and in fact improves it
+slightly: if the cluster is partitioned into two even halves then one of the
+halves will contain a majority of the masters and will be able to keep
+operating, whereas if all of the master-eligible nodes' votes were counted then
+neither side could make any progress in this situation.
+
+[float]
+==== Adding master-eligible nodes
+
+If you wish to add some master-eligible nodes to your cluster, simply configure
+the new nodes to find the existing cluster and start them up. Once the new
+nodes have joined the cluster, you may be able to increase the
+`cluster.master_nodes_failure_tolerance` setting to match. (TODO do we log
+info/warn messages about this?)
+
+[float]
+==== Removing master-eligible nodes
+
+If you wish to remove some of the master-eligible nodes in your cluster, you
+must first reduce the `cluster.master_nodes_failure_tolerance` setting to match
+the target cluster size before removing the extraneous nodes. This temporary
+situation is the only case in which `cluster.master_nodes_failure_tolerance`
+should be set lower than the recommended values above.
+
+You must also be careful not to remove too many master-eligible nodes all at
+the same time. For instance, if you currently have seven master-eligible nodes
+and you wish to reduce this to three, you cannot simply stop four of the nodes
+all at the same time: to do so would leave only three nodes remaining, which is
+less than half of the cluster, which means it cannot take any further actions.
+You should remove the nodes one-at-a-time and verify that each node has been
+removed from the configuration before moving onto the next one, using the
+`await_removal` API:
+
+[source,js]
+--------------------------------------------------
+# Explicit timeout of one minute
+GET /_nodes/node_name/await_removal?timeout=1m
+# Default timeout of 30 seconds
+GET /_nodes/node_name/await_removal
```

Comment:
Maybe we can combine it with cluster health, which already has a `wait_for_nodes` condition where you can specify a count. Maybe we can add a `wait_for_master_eligible_nodes` condition to the cluster health API, combined with a `wait_for_optimal_configuration` condition?

Reply:
Yes, I'm not sure that resolving the node using `DiscoveryNodes#resolveNodes()` is the right thing to do here because we could be talking about nodes that no longer exist. On the one hand, failing to resolve the node means it's not in the configuration, which should count as success. On the other hand it's possible that this would be trappy for at least two reasons:

- if you ask for a node that never existed by mistake then this will indicate immediate success, which could count as undue leniency.
-  if removing a node causes a different node to match the spec instead then this might return a false negative.

I think to solve these we'd need to keep master-node tombstones around. Even if we only accepted the persistent node ID I think we would still need to distinguish "this node is no longer here" from "this node never existed" in our response.

This API is separate from the retirement API because retirement changes state and requires cleanup (I should spell this out in more detail) whereas this is purely a wait-for-state check. It's possible we could extend the health API, but e.g. when going from 4 nodes to 3 you need to care about _which_ 3 nodes are still in the configuration, and not just that there are 3 of them.

Reply:
Note that if retirement API were only to be used for the 2 to 1 node case, we would not actually need to store any persistent data about retired nodes in the cluster state. It would be sufficient to shift the vote to the correct node. I feel that we have one API too much here.

Reply:
I'm contemplating an alternative to the wait API in which you can ask "is it safe to shut down nodes _{X,Y,Z}_ (while maintaining fault tolerance _n_)" since that is, ultimately, the question you want to ask when doing orchestration.

I also see value in the retirement API for whole-cluster migrations (e.g. 3->6->3) in which you could retire the old 3 nodes once the new 3 are up, although I didn't include this in the document.

Reply:
I have removed this API in favour of retirement (or just "wait long enough") in the latest version.